velox/functions/prestosql/ArrayMinMaxBy.cpp

```diff
@@ -0,0 +1,179 @@
+ * limitations under the License.
+ */
+
+#include <folly/container/F14Set.h>
+
+#include <glog/logging.h>
+#include <velox/type/SimpleFunctionApi.h>
+#include "velox/expression/EvalCtx.h"
+#include "velox/expression/Expr.h"
+#include "velox/expression/VectorFunction.h"
+#include "velox/functions/lib/LambdaFunctionUtil.h"
+#include "velox/functions/lib/RowsTranslationUtil.h"
+
+namespace facebook::velox::functions {
+namespace {
+
+template <bool isMaxBy>
+class ArrayMinMaxByFunction : public exec::VectorFunction {
+ public:
+  void apply(
+      const SelectivityVector& rows,
+      std::vector<VectorPtr>& args,
+      const TypePtr& outputType,
+      exec::EvalCtx& context,
+      VectorPtr& result) const override {
+    VELOX_CHECK_EQ(args.size(), 2);
+    context.ensureWritable(rows, outputType, result);
+
+    exec::LocalDecodedVector arrayDecoder(context, *args[0], rows);
+    auto& decodedArray = *arrayDecoder.get();
+
+    // Flatten input array.
+    auto flatArray = flattenArray(rows, args[0], decodedArray);
+
+    std::vector<VectorPtr> lambdaArgs = {flatArray->elements()};
+    auto newNumElements = flatArray->elements()->size();
+
+    SelectivityVector validRowsInReusedResult =
+        toElementRows<ArrayVector>(newNumElements, rows, flatArray.get());
+
+    VectorPtr newElements;
+    auto elementToTopLevelRows = getElementToTopLevelRows(
+        newNumElements, rows, flatArray.get(), context.pool());
+
+    // Loop over lambda functions and apply these to elements of the base array.
+    // In most cases there will be only one function and the loop will run once.
+    auto it = args[1]->asUnchecked<FunctionVector>()->iterator(&rows);
+    while (auto entry = it.next()) {
+      auto elementRows = toElementRows<ArrayVector>(
+          newNumElements, *entry.rows, flatArray.get());
+      auto wrapCapture = toWrapCapture<ArrayVector>(
+          newNumElements, entry.callable, *entry.rows, flatArray);
+
+      entry.callable->apply(
+          elementRows,
+          &validRowsInReusedResult,
+          wrapCapture,
+          &context,
+          lambdaArgs,
+          elementToTopLevelRows,
+          &newElements);
+    }
+
+    // New elements after the lambda function is applied
+    exec::LocalDecodedVector decodedElements(
+        context, *newElements, validRowsInReusedResult);
+
+    BufferPtr resultIndices = allocateIndices(rows.size(), context.pool());
+    auto* rawResultIndices = resultIndices->asMutable<vector_size_t>();
+
+    BufferPtr resultNulls = allocateNulls(rows.size(), context.pool());
+    auto* mutableNulls = resultNulls->asMutable<uint64_t>();
+
+    auto indices = getMaxOrMinIndices(rows, *flatArray, *newElements, context);
+    rows.applyToSelected([&](auto row) {
+      auto offset = flatArray->offsetAt(row);
+      if (decodedElements->isNullAt(offset)) {
+        bits::setNull(mutableNulls, row);
+      }
+      rawResultIndices[row] = indices[row];
```

Comment:
Why not just populate rawResultIndices inside getMaxOrMinIndices, then if my suggestion above applies, you can just get rid of this whole applyToSelected